File: src/include/fetch/FetchedWebItem.php
<?php

namespace fetch;

use DateTime;
use DOMDocument;
use DOMNodeList;
use DOMXPath;
use SimpleXMLElement;

class FetchedWebItem
{

	public function __construct(
		public readonly WebItemReference $reference,
		public readonly ?string $content,
		public readonly ?int $httpStatusCode,
		public readonly ?DateTime $lastUpdated
	) {}

	/**
	 * 
	 * @return WebItemReference[]
	 */
	public function parseAsSitemap(?DateTime $filterNewerThan = NULL): ?array
	{

		$xml = $this->parseAsSimpleXmlElement();
		if ($xml === null) {
			return null;
		}

		$urls = [];
		foreach ($xml->url as $urlEntry) {
			$loc = (string)$urlEntry->loc;
			$lastmod = isset($urlEntry->lastmod) ? new DateTime((string)$urlEntry->lastmod) : new DateTime();
			if ($filterNewerThan === null || $lastmod >= $filterNewerThan) {
				$urls[] = new WebItemReference(url: $loc);
			}
		}

		return $urls;
	}

	/**
	 * 
	 * @return array<string> 
	 */
	public function parseAsCommaSeparatedList(): array
	{
		if ($this->content === null) {
			return [];
		}

		$content = str_replace("\n", "", $this->content);

		$contentArray = explode(",", $content);
		$contentArray = array_filter($contentArray);
		$contentArray = array_map('trim', $contentArray);

		return $contentArray;
	}

	/**
	 * 
	 * @return null|array<mixed,mixed>
	 */
	public function parseAsJson(): ?array
	{
		if ($this->content === null) {
			return null;
		}
		$result = json_decode($this->content, associative: true);

		return $result;
	}

	public function parseAsSimpleXmlElement(): ?SimpleXMLElement
	{
		if ($this->content === null) {
			return null;
		}
		$result = new SimpleXMLElement($this->content);

		return $result;
	}

	public function parseAsDomDocument(): ?DOMDocument
	{
		if ($this->content === null) {
			return null;
		}
		$document = new DOMDocument();
		@$document->loadHTML($this->content);
		return $document;
	}

	/**
	 * 
	 * @return null|array<string,string>
	 */
	public function parseHtmlMetaTags(): ?array
	{
		$output = [];
		$document = $this->parseAsDomDocument();
		if ($document === null) {
			return null;
		}

		$metaTags = $document->getElementsByTagName('meta');
		foreach ($metaTags as $tag) {
			if ($tag->getAttribute('name') != "") {
				$output[$tag->getAttribute('name')] = $tag->getAttribute('content');
			} elseif ($tag->getAttribute('property')  != "") {
				$output[$tag->getAttribute('property')] = $tag->getAttribute('content');
			}
		}
		return $output;
	}

	/*public function parseAsWordpressApiPosts()
	{
		// Not implemented yet
	}*/
}

File: app/fetch/fetched_web_item.py
from __future__ import annotations

import io
import json
from dataclasses import dataclass
from typing import Optional, TYPE_CHECKING
from xml.etree import ElementTree

from bs4 import BeautifulSoup
from PIL import Image

if TYPE_CHECKING:
    from app.fetch.web_item_reference import WebItemReference


@dataclass
class FetchedWebItem:
    reference: "WebItemReference"
    content: Optional[bytes]
    http_status_code: Optional[int]

    def parse_as_json(self) -> Optional[dict | list]:
        if self.content is None:
            return None
        try:
            return json.loads(self.content.decode("utf-8"))
        except (json.JSONDecodeError, UnicodeDecodeError):
            return None

    def parse_as_sitemap(self, filter_newer_than=None) -> Optional[list["WebItemReference"]]:
        from app.fetch.web_item_reference import WebItemReference
        from datetime import datetime, timezone

        if self.content is None:
            return None
        try:
            root = ElementTree.fromstring(self.content)
            ns = {"sm": "http://www.sitemaps.org/schemas/sitemap/0.9"}
            urls: list[WebItemReference] = []
            for url_el in root.findall("sm:url", ns) or root.findall("url"):
                loc_el = url_el.find("sm:loc", ns) or url_el.find("loc")
                if loc_el is None or loc_el.text is None:
                    continue
                loc = loc_el.text.strip()

                if filter_newer_than is not None:
                    lastmod_el = url_el.find("sm:lastmod", ns) or url_el.find("lastmod")
                    if lastmod_el is not None and lastmod_el.text:
                        try:
                            lastmod = datetime.fromisoformat(lastmod_el.text.strip())
                            if lastmod.tzinfo is None:
                                lastmod = lastmod.replace(tzinfo=timezone.utc)
                            if filter_newer_than.tzinfo is None:
                                filter_newer_than = filter_newer_than.replace(tzinfo=timezone.utc)
                            if lastmod < filter_newer_than:
                                continue
                        except ValueError:
                            pass

                urls.append(WebItemReference(url=loc))
            return urls
        except ElementTree.ParseError:
            return None

    def parse_as_xml_element(self) -> Optional[ElementTree.Element]:
        if self.content is None:
            return None
        try:
            return ElementTree.fromstring(self.content)
        except ElementTree.ParseError:
            return None

    def parse_as_comma_separated_list(self) -> list[str]:
        if self.content is None:
            return []
        text = self.content.decode("utf-8", errors="replace").replace("\n", "").replace("\r", "")
        parts = [p.strip() for p in text.split(",") if p.strip()]
        return parts

    def parse_as_dom(self) -> Optional[BeautifulSoup]:
        if self.content is None:
            return None
        return BeautifulSoup(self.content, "lxml")

    def parse_html_meta_tags(self) -> Optional[dict[str, str]]:
        soup = self.parse_as_dom()
        if soup is None:
            return None
        output: dict[str, str] = {}
        for tag in soup.find_all("meta"):
            name = tag.get("name") or tag.get("property") or ""
            content = tag.get("content", "")
            if name:
                output[name] = content
        return output

    def parse_as_image(self) -> Optional[Image.Image]:
        if self.content is None:
            return None
        try:
            img = Image.open(io.BytesIO(self.content))
            img.load()  # Force decode
            return img.convert("RGBA")
        except Exception:
            return None
